File: src/infrastructure/storage/project_repository.rs
use crate::domain::entities::project::ProjectInfo;
use chrono::{DateTime, Local};
use directories::ProjectDirs;
use rusqlite::{params, Connection, Result as SqlResult};
use std::fs;
use std::io;
use std::path::PathBuf;

pub struct ProjectRepository {
    conn: Connection,
}

impl ProjectRepository {
    pub fn new() -> SqlResult<Self> {
        let db_path = Self::get_database_path()
            .map_err(|e| rusqlite::Error::InvalidPath(e.to_string().into()))?;

        // Verificar si el archivo existe
        let db_exists = db_path.exists();

        // Crear conexión
        let conn = Connection::open(&db_path)?;

        // Si es una nueva base de datos, inicializarla
        if !db_exists {
            Self::initialize_database(&conn)?;
        }

        Ok(Self { conn })
    }

    fn initialize_database(conn: &Connection) -> SqlResult<()> {
        // Crear tabla de proyectos
        conn.execute(
            "CREATE TABLE IF NOT EXISTS recent_projects (
                id INTEGER PRIMARY KEY,
                name TEXT NOT NULL,
                path TEXT NOT NULL UNIQUE,
                created_at TEXT NOT NULL,
                last_opened TEXT NOT NULL
            )",
            [],
        )?;

        // Crear índices
        conn.execute(
            "CREATE INDEX IF NOT EXISTS idx_last_opened 
             ON recent_projects(last_opened DESC)",
            [],
        )?;

        conn.execute(
            "CREATE UNIQUE INDEX IF NOT EXISTS idx_path 
             ON recent_projects(path)",
            [],
        )?;

        Ok(())
    }

    fn get_database_path() -> io::Result<PathBuf> {
        let proj_dirs = ProjectDirs::from("com", "cortex", "engine").ok_or_else(|| {
            io::Error::new(
                io::ErrorKind::NotFound,
                "No se pudo obtener directorio de datos",
            )
        })?;

        let data_dir = proj_dirs.data_dir();
        fs::create_dir_all(data_dir)?;

        Ok(data_dir.join("projects.db"))
    }

    pub fn add_project(&self, project: &ProjectInfo) -> SqlResult<()> {
        self.conn.execute(
            "INSERT OR REPLACE INTO recent_projects (name, path, created_at, last_opened)
             VALUES (?1, ?2, ?3, ?4)",
            params![
                project.name,
                project.path.to_string_lossy().to_string(),
                project.created_at.to_rfc3339(),
                Local::now().to_rfc3339(),
            ],
        )?;
        Ok(())
    }

    pub fn get_recent_projects(&self, limit: i64) -> SqlResult<Vec<ProjectInfo>> {
        let mut stmt = self.conn.prepare(
            "SELECT name, path, created_at 
             FROM recent_projects 
             ORDER BY last_opened DESC 
             LIMIT ?",
        )?;

        let projects = stmt.query_map([limit], |row| {
            Ok(ProjectInfo {
                name: row.get(0)?,
                path: PathBuf::from(row.get::<_, String>(1)?),
                created_at: DateTime::parse_from_rfc3339(&row.get::<_, String>(2)?)
                    .unwrap()
                    .with_timezone(&Local),
            })
        })?;

        projects.collect()
    }

    pub fn update_last_opened(&self, path: &PathBuf) -> SqlResult<()> {
        self.conn.execute(
            "UPDATE recent_projects 
             SET last_opened = ? 
             WHERE path = ?",
            params![
                Local::now().to_rfc3339(),
                path.to_string_lossy().to_string(),
            ],
        )?;
        Ok(())
    }

    pub fn remove_project(&self, path: &PathBuf) -> SqlResult<()> {
        self.conn.execute(
            "DELETE FROM recent_projects WHERE path = ?",
            params![path.to_string_lossy().to_string()],
        )?;
        Ok(())
    }

    pub fn exists(&self, path: &PathBuf) -> SqlResult<bool> {
        let count: i64 = self.conn.query_row(
            "SELECT COUNT(*) FROM recent_projects WHERE path = ?",
            params![path.to_string_lossy().to_string()],
            |row| row.get(0),
        )?;

        Ok(count > 0)
    }
}

File: src/infrastructure/storage/ProjectRepository.cpp
#include "ProjectRepository.hpp"

namespace Infrastructure
{
    ProjectRepository::ProjectRepository()
    {
        dbPath = getDatabasePath();
        initializeDatabase();
    }

    ProjectRepository::~ProjectRepository()
    {
        // Implementación del destructor
    }

    void ProjectRepository::initializeDatabase()
    {
        int rc = sqlite3_open(dbPath.string().c_str(), &db);
        if (rc)
        {
            throw std::runtime_error("No se pudo abrir la base de datos");
        }

        const char *sql = R"(
            CREATE TABLE IF NOT EXISTS projects (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL,
                path TEXT UNIQUE NOT NULL,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                last_opened DATETIME DEFAULT CURRENT_TIMESTAMP
            );
        )";

        char *errMsg = nullptr;
        rc = sqlite3_exec(db, sql, nullptr, nullptr, &errMsg);
        if (rc != SQLITE_OK)
        {
            std::string error = errMsg;
            sqlite3_free(errMsg);
            throw std::runtime_error("Error al crear tabla: " + error);
        }
    }

    bool ProjectRepository::addProject(const Domain::ProjectInfo &project)
    {
        const char *sql = R"(
            INSERT OR REPLACE INTO projects (name, path, last_opened)
            VALUES (?, ?, CURRENT_TIMESTAMP);
        )";

        sqlite3_stmt *stmt;
        int rc = sqlite3_prepare_v2(db, sql, -1, &stmt, nullptr);
        if (rc != SQLITE_OK)
            return false;

        sqlite3_bind_text(stmt, 1, project.getName().c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(stmt, 2, project.getPath().string().c_str(), -1, SQLITE_STATIC);

        rc = sqlite3_step(stmt);
        sqlite3_finalize(stmt);

        return rc == SQLITE_DONE;
    }

    std::vector<Domain::ProjectInfo> ProjectRepository::getRecentProjects(int limit)
    {
        std::vector<Domain::ProjectInfo> projects;
        const char *sql = R"(
            SELECT name, path FROM projects 
            ORDER BY last_opened DESC LIMIT ?;
        )";

        sqlite3_stmt *stmt;
        int rc = sqlite3_prepare_v2(db, sql, -1, &stmt, nullptr);
        if (rc != SQLITE_OK)
            return projects;

        sqlite3_bind_int(stmt, 1, limit);

        while (sqlite3_step(stmt) == SQLITE_ROW)
        {
            std::string name = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 0));
            std::string path = reinterpret_cast<const char *>(sqlite3_column_text(stmt, 1));
            projects.emplace_back(name, path);
        }

        sqlite3_finalize(stmt);
        return projects;
    }

    std::filesystem::path ProjectRepository::getDatabasePath()
    {
        std::filesystem::path userHome = std::getenv("HOME") ? std::getenv("HOME") : std::getenv("USERPROFILE");
        std::filesystem::path dbDir = userHome / ".cortex";

        // Crear el directorio si no existe
        if (!std::filesystem::exists(dbDir))
        {
            std::filesystem::create_directory(dbDir);
        }

        return dbDir / "projects.db";
    }
}